moveit_core/robot_state/src/robot_state.cpp

```diff
@@ -985,55 +987,111 @@ bool RobotState::clearAttachedBody(const std::string& id)
     return false;
 }
 
-const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& id)
+const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& frame_id)
 {
   updateLinkTransforms();
-  return static_cast<const RobotState*>(this)->getFrameTransform(id);
+  return static_cast<const RobotState*>(this)->getFrameTransform(frame_id);
 }
 
-const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& id) const
+const Eigen::Isometry3d& RobotState::getFrameTransform(const std::string& frame_id) const
 {
-  if (!id.empty() && id[0] == '/')
-    return getFrameTransform(id.substr(1));
+  static Eigen::Isometry3d transform;  // TODO(felixvd): Is this safe??
```

Comment:
@felixvd this static variable is shared across all calls to this function, right? So if someone does:
```
const Eigen::Isometry3d& transform1 = robot_state.getFrameTransform("foo");
const Eigen::Isometry3d& transform2 = robot_state.getFrameTransform("bar");
```
Then After those two lines of code, transform1 will have the transform for "bar", right? (they both will?) This makes sense but seems icky.....

Reply:
That's my understanding, so I can't imagine this is right. What is the right way to do this though? A `unique_ptr`? Even though the return type is a reference? I don't know enough C++ to get through this efficiently on my own, so I'm throwing this out there.

Reply:
I think this is a limation of returning a reference. It works OK if there's a semi-permanent object in the class that you can return a reference to. It doesn't work for computed-on-the-fly values, since there is no long term value in memory to return a reference to.

I can see two options:
1. keep an `std::map` of transforms for every frame within an object. Return references to those transforms. Optionally, you could still do lazy computation of the transform, keeping a "ditry" marker or something, and only computing them as needed.
2. Switch to return-by-value for these transforms

Reply:
I actually lean more towards option (2). Returning a reference seems like a bit of optimization overkill for such a small object.

Reply:
There's a high chance that compilers used will use RVO here anyway, so just returning a normal object is probably OK.

Reply:
The original version of these functions already returned references(`const Eigen::Isometry3d& getFrameTransform(const std::string frame_id)'), so this change is not for optimization. It is to return the results of `knowsTransform` and `getFrameTransform` in a single function. I could overload the original getFrameTransform to `const Eigen::Isometry3d& getFrameTransform(const std::string frame_id, const bool transform_successfully_found)` or something, but this pattern made more sense to me. I can't point to a guideline, but I assumed it is good practice, since you can write compact lines like [this](https://github.com/ros-planning/moveit/pull/1439/files#diff-e03d4d985fd8f5e9565368bddcac8d96R309).

If there's no way to implement this cleanly then I'll make the new functions return the references too, but maybe someone has an idea.

@jonbinney For the record, the references in the "main" `getFrameTransform`/`getFrameInfo` function are set to objects in the RobotState that outlive the function. It's only in the wrapper functions that an intermediate object is created so it can be returned.

Reply:
Two more thoughts, after I posted [an SO question](https://stackoverflow.com/questions/55756238/return-a-reference-from-a-function-that-takes-it-as-a-parameter) about this:

1) I think the fact that the reference is non-const was unsafe to start with, because it allows values in the `RobotState` to be changed.

2) Apparently C++ 17 introduced [std::optional](https://en.cppreference.com/w/cpp/utility/optional) for cases like this. I will try that out.

Reply:
Done in https://github.com/ros-planning/moveit/pull/1439/commits/4915f38946acbe526e4011aee721f9d273dae890, so I am resolving this.

Reply:
Thanks for the input!